tools/scripts/gen-attributions.py

```diff
@@ -0,0 +1,253 @@
+        text: str,
+        name: Optional[str] = None,
+        type: Optional[str] = None,
+        master: str = "false",
+        agent: str = "false",
+        webui: str = "false",
+    ) -> None:
+        assert text.strip(), "License text not found"
+        assert name is not None, "a Name header is required"
+        assert type is not None, "a Type header is required"
+        assert (
+            type.lower() in known_licenses
+        ), f"Type header must be one of {known_licenses}"
+        assert master.lower() in {"true", "false"}, "Master must be true or false"
+        assert agent.lower() in {"true", "false"}, "Agent must be true or false"
+        assert webui.lower() in {"true", "false"}, "Webui must be true or false"
+
+        self.tag = tag
+        self.text = text
+        self.name = name
+        self.type = type.lower()
+        self.master = master.lower() == "true"
+        self.agent = agent.lower() == "true"
+        self.webui = webui.lower() == "true"
+
+    @classmethod
+    def from_file(cls, tag: str, f: IO[str]) -> "License":
+        # Licenses are saved as a simple rfc822 format (email headers + body).
+        msg = email.message_from_file(f)
+        meta = {x[0].lower(): x[1] for x in msg._headers}  # type: ignore
+        text = msg.get_payload()
+        return cls(tag, text, **meta)
+
+    def sphinx_ref(self) -> str:
+        """
+        Example output:
+            :ref:`BSD 3-clause <tomb>`
+        """
+        return f":ref:`{known_licenses[self.type]} <{self.tag}>`"
+
+    def sphinx_entry(self) -> str:
+        return "\n".join(
+            [
+                f".. _{self.tag}:",
+                "",
+                sphinx_format_header(self.name, "*"),
+                "",
+                ".. code::",
+                "",
+                indent(self.text, "   "),
+            ]
+        )
+
+    def ascii_entry(self) -> str:
+        return "\n".join(
+            [
+                f"{self.name}",
+                "",
+                indent(self.text, "    "),
+            ]
+        )
+
+
+def sphinx_format_header(text: str, char: str) -> str:
+    """
+    Example output:
+
+        *******
+         WebUI
+        *******
+    """
+    return "\n".join(
+        [
+            char * (len(text) + 2),
+            f" {text}",
+            char * (len(text) + 2),
+        ]
+    )
+
+
+def gen_sphinx_table(licenses: List[License]) -> str:
+    """
+    Example output:
+
+        .. list-table::
+           :header-rows: 1
+
+           * - Package
+             - License
+           * - gopkg.in/tomb.v1
+             - :ref:`BSD 3-clause <tomb>`
+    """
+
+    lines = [
+        ".. list-table::",
+        "   :header-rows: 1",
+        "",
+        "   * - Package",
+        "     - License",
+    ]
+
+    for license in licenses:
+        lines.append(f"   * - {license.name}")
+        lines.append(f"     - {license.sphinx_ref()}")
+
+    return "\n".join(lines)
```

Comment:
non-blocking: I don't quite think this is actually worth doing (though feel free to if you feel excited about it), but, just to point out an interesting technique, it could be elegant to switch everything to `yield`ing lines instead, which would save a lot of `foo.splitlines()` and `"\n".join(foo)` and so on. (I guess you might have to put a few new ones in to handle the existing chunks of text, like the preamble below, but it would probably still be a win overall.)

Reply:
That's a neat idea.  You could yield lines and instead of joining chunks of text, you could use itertools to chain together iterators (although you'd have to write your own joining routine maybe).

But I think I like it how it is.  There's only two places where `splitlines()` is used: `indent` and `post_process`, which is pretty managable.  In general, you can think of those as the only exceptions to the rule that "all functions return blocks of text which can be joined with other blocks to create larger blocks of text".  And I think that's just a lot easier to think about that chaining generators together... especially when it all runs in like 50 ms anyway : )